doc/cli_subprocess_guidelines.md

```diff
@@ -0,0 +1,110 @@
+
+There are several aspects of security practices that developers need to have in mindset to safeguard their cli modules from command injection attacks.
+
+### Cli Centralized Subsystem Executing
+
+Azure cli provides a centralized function `run_cmd` adapted from official `subprocess.run`, with necessary argument covered and illegal input blocking enforced. 
+
+What developers need to do is:
+1) `from azure.cli.core.util import run_cmd`
+2) replace `subprocess.run` (or `Popen` or `check_call` or `check_output` or `call`) with `run_cmd`.
+3) construct cmd args as array like: [executable, arg0, arg1, arg2, ...]
+
+`run_cmd` will add necessary argument type checks and process the input and output the same way as `subprocess.run`, and block potential risks from commands constructed from user input.
+Below is an example for `run_cmd` use case:
+
+```commandline
+# code before:
+cmd = f"git commit -m {user_message}"
+import subprocess
+output = subprocess.run(cmd, shell=True, check=False, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
+if output.returncode != 0:
+      raise CLIInternalError('Command execution failed, command is: '{}', error message is: {}'.format(cmd, output.stderr))
+return output.stdout
+```
+
+If `run_cmd` is applied, it would be like:
+```commandline
+# code after:
+cmd = ["git","commit", "-m", user_message]
+import subprocess
+from azure.cli.core.util import run_cmd
+output = run_cmd(cmd, check=False, capture_output=True)
+if output.returncode != 0:
+      raise CLIInternalError('Command execution failed, command is: '{}', error message is: {}'.format(cmd, output.stderr))
+return output.stdout
+```
+
+All various kinds of `subprocess` Popen calling use cases can be easily adjusted into `run_cmd` with security risks processed and eliminated in this centralized function.
+
+Besides that, users might need to know some parts of the accessibility in both `run_cmd` and `subprocess`
+1) when calling shell built-in cmds, like `dir` or `echo`, using `shell=True` **in windows platform**, `subprocess` implicitly uses `cmd.exe`, while `run_cmd` asks developers to provide the `cmd.exe` as executable file specifically in the arg list's first item, like `["cmd.exe", "/c", "echo", "abc"]`
+2) if developers want to find an easy way to split their current cmd string into list, **for unix-like platforms**, developers can apply [`shlex.split`](https://docs.python.org/3/library/shlex.html#shlex.split) for quick access. But a prepared cmd statement is still more recommended (for more info about prepared cmd statement, please read below sections).
+3) it might be not that obvious to find target command's executable file **in windows platform**, a tool developer can use is `shutil.which` that gives the executable file path in windows system, like `shutil.which(git)`. The cmd `git --version` can be adjusted as `[shutil.which(git), "--version"]`. Please provide the corresponding executable path in target platforms.
```

Comment:
Could you share more information on this? I think `subprocess` will look for the executable on `PATH`.

Reply:
Yes, subprocess searches executable under `PATH`, as the same way as `shutil.which`. 

In windows platform, cmd-running can be `cmd arg0 arg2..` or `cmd.exe arg0 arg1...`. Just want to give a hint to users how to check the absolute path of corresponding executable file in case they are confused with it. 

Doc adjusted.